### ISSUE TITLE
deploy of example failing with an auth error

### ISSUE THREAD

**Issue by kcoleman-marklogic** [marklogic/entity-services#198]
This will probably turn out to be user error. I have a clean installation of the server and a clone of the project. I just want to run the examples, so I went directly to the examples dir to try to do so. I did not run gradle at all in parent dir.

I modified the gradle.properties in examples dir to change mlHost (mine is remote), mlAdminPassword, mlRestAdminPassword, and mlManagePassword.

I then ran `./gradlew --info -PexampleDir=example-person mlDeploy`

And it dies with an auth failure:

```
Executing command [com.marklogic.appdeployer.command.databases.DeploySchemasDatabaseCommand] with sort order [100]
Sending XML GET request as user 'adam' to path: /manage/v2/databases
GET request for "http://kim-2.marklogic.com:8002/manage/v2/databases" resulted in 401 (Unauthorized); invoking error handler
Logging HTTP response body to assist with debugging: {"errorResponse":   {"statusCode":401,
   "status":"Unauthorized",
   "message":"401 Unauthorized"
  }
}
```

I tried the HTTP request manually, and it does indeed fail. What's peculiar to me is that it appeared as if several other mgmt REST API commands had already successfully run as the "adam" user. For example, I see this in the log:

```
Sending XML GET request as user with admin role 'adam' to path: /manage/v2/users
Found user with name of whitney, so updating at path /manage/v2/users/whitney/properties
```

However, you can't run GET /manage/v2/users now. At least not with the password I specified in gradle.properties.

I'm assuming the problem is the password, rather than the privs. Was I not supposed to set the password? I assumed I was since it was "********" in the properties file.

When I look at the user conf, it has the `race-admin` role, which in turn has `manage-admin`, `race-writer`, and `rest-admin`. Seems reasonable (though IDK what it should be), which reinforces my assumption the password might be the problem.

I will keep digging at this, but wanted to go ahead and file it in case you have insights to contribute.

**Comment by kcoleman-marklogic** [comment]
Editing entity-services-examples/gradle.properties to set `mlRestAdminPassword` and `mlManagePassword` back to a series of asterisks made this problem go away. Is that WAI?

My guess it maybe changing the passwords in gradle.properties is conflicting with what's in entity-services-examples/src/main/ml-config/security/users/adam.json.

(Why, oh, why can't we just call that directory "examples". Lol.)

**Comment by grechaw** [comment]
Some EA folks got the same issue.  srikanth tracked it down to a bug in ml-gradle around deploy/undeploy sequencing.

**Comment by grechaw** [comment]
setting to fix -- we need the examples to be more fool-proof.

**Comment by bsrikan** [comment]
I think what Kim mentioned here is different from what EA folks were doing although the error message is same. Kim has tried deploying from the examples dir and she found the root case to be setting of password to something other than "********"

**Comment by grechaw** [comment]
Just to collect more issues here....
If you run gradle deploy from parent directory, it does not see schemas config for children (which are relative to the child directories.  So current state is that you have to run gradle mldeploy from subdirectories for unit tests all to pass, too.

**Comment by grechaw** [comment]
I'm experimenting with configuration of security.  
One way to make the deployment more idiot-proof (we think @bsrikan found a bug in using highly-privileged users who are not admin in ml-gradle), would be to use an admin user for administration, but 'user' users for running tests and apps.

@jmakeig did good work in putting security into the examples, but having done so he complexified the configuration of these multiple projects.

**Comment by grechaw** [comment]
For these projects, we have three levels of admin -- the 'admin' user has to create and remove users.  This user is currently only set at the parent project level.

I'm going to suggest that we dispense with the third 'middle admin' role.  It's causing issues for us that will make adoption harder, and yet most of the time, admin/admin will indeed be the administrative credentials for folks working with entity services, because it's a development tool.

**Comment by jmakeig** [comment]
Is the issue the security configuration or the multiple projects? I hadn’t considered the latter when I implemented what’s there. (And that’s probably more of a consideration for our demo potpourri than someone configuring a single application for herself.) Regardless, we need to get to the point where the set-up “just works”.

**Comment by kcoleman-marklogic** [comment]
Perhaps some of this is handable through documentation. What you must/should/can do and what you must not do in standing up the examples. This ties back into the task I filed the other day. See #199.

**Comment by grechaw** [comment]
Correct me if I'm mistaken @bsrikan but I think we're hitting an actual bug in ml-gradle.
The undeploy command, somewhere in the interaction of multiproject and ml-gradle, removes the rest-admin/manage user before trying to use him (adam) to remove things.  I think @bsrikan was going to file something against ml-gradle. 

In my tests right now, when I just set tall the mlAdminUsername to admin, then the whole combination of deploy and undeploy seems to work from the parent project (and the child ones).

**Comment by rjrudin** [comment]
Where does it fail? Users and roles aren't undeployed until near the very end. See https://github.com/rjrudin/ml-app-deployer/blob/master/src/main/java/com/marklogic/appdeployer/command/SortOrderConstants.java#L62 - the higher the number, the later it runs during mlUndeploy. So users, then roles, then mimetypes, then groups, then done. 

Now - if this is multi-project, and Project A defines a user that Project B uses, and Project A is undeployed before Project B - then Project B will fail to undeploy because it'll try to use a user that has been deleted. Project B would need to be undeployed first.

**Comment by bsrikan** [comment]
What you mentioned Charles about users getting removed first before databases are removed is what I observed as well.

For some reason now undeploy for e2e(not a multiproject) and examples(multiproject) didnt give any authorization error. I am going to try again from scratch ie deploy run test then undeploy and report if I get the auth error.

**Comment by kcoleman-marklogic** [comment]
You guys probably already understand this, but just to be certain: The auth error I got was not during an undeploy. It was while deploying the example. No "un".

**Comment by bsrikan** [comment]
Yes Kim. Moreover you tried setting a password other than "********" and you werent getting the auth error once you set the *s back in the password field right.

I think we are just listing all auth errors that we encounter while running examples here.

**Comment by bsrikan** [comment]
I dont get auth error on deploy or undeploy for examples when running from examples dir. 

We would get auth error on deploy or undeploy when we try to do so from parent project because the users are config to setup only in examples(and e2e)sub-projects. I think we did chat about this Charles that we would not want to move the ml-config/security to parent dir.

Except races none of other projects in examples have any user/role or security setup.

**Comment by grechaw** [comment]
As an experiment though I did remove the admin users for e2e and examples.
deploy and undeploy are faster now (removing the unused roles would speed it up more)
I kind of like the improved UX of making the deploys faster.

Do you all support simplifying the security model if it is better UX?  It's just less likely to end up in some tricky unrecoverable state, even if there's no real bug.

**Comment by kcoleman-marklogic** [comment]
Less is more, IMO. The focus should be on ES, anyway. This isn't SampleStack.

**Comment by grechaw** [comment]
As the auth for examples was never something specified or required, I'd like to proceed with the simplificaftion.  @jmakeig one final OK from you?  I'll put the change into a branch momentarily for review.

**Comment by jmakeig** [comment]
As long as it doesn't mean doing non-admin things as the `admin` role, like loading or transforming data.

**Comment by grechaw** [comment]
Yes, that's a good point.  we are indeed covered there, with a runtime user who does the work in the java layer (harmonize, load, etc)

**Comment by grechaw** [comment]
PR #205 addresses this issue with simplification of security model.

**Comment by grechaw** [comment]
This bug is not exactly what was fixed with the recent PR merge, but the PR does take care of auth issues in general.  Do you want to give it a whirl now and see if it works better?  I'm going to also update the NPE error you were seeing in issue #201

**Comment by kcoleman-marklogic** [comment]
This is still not working for me. I still get an auth failure, though at a different stage of the deployment. It's getting a 401 when doing a `GET /manage/v2/roles`. A plain ol' curl request with the correct password works.

Do you want to track it here, or would you like me to file a new issue?

**Comment by kcoleman-marklogic** [comment]
P.S. Working with today's server build, a fresh build of the Java Client (which appears to be the origin of the rejected request, based on the debug log), and the latest ES sources.

**Comment by grechaw** [comment]
I need steps to reproduce.  The instructions for deploying examples have changed.  I suspect that you don't do things in an OOTB way (like, heaven forbid, not using admin/admin or something).  So I think you've got a valid issue, I just don't know what it is.

Important for examples are running from the examples dir, and using the -PexampleDir parameter.

**Comment by grechaw** [comment]
heh FWIW when I just tried to repro I got a 500 error and missing module (need to rebuild)

**Comment by kcoleman-marklogic** [comment]
Steps were the same as in the original filing of the issue.
* Clone the ES project.
* Went into the entity-services-examples directory
* Modified gradle.properties to change `mlHost` and `mlAdminPassword`
* Ran `./gradlew -PexampleDir=example-person mlDeploy`

You say the instructions have changed, but when I look at the README, it still looks like the above is the correct deploy command. What's the correct way to do it now?

**Comment by kcoleman-marklogic** [comment]
And no, I don't have an OOTB configuration. My ML host is remote and my admin password is not "admin". I think the latter is the likely source of difficulty, just like before.

**Comment by grechaw** [comment]
That looks right (they're different from EA-3 is what I meant).
maybe I have some boneheaded bug involving admin password -- i remember such a bug from samplestack days

**Comment by grechaw** [comment]
Not sure what the problem is, but I reproduced just now indirectly by changing the password in the properties file.  mldeploy still succeeded.  So the credentials for ml-gradle are not matching those in the properties file.  Maybe it's just the wrong property name, or perhaps multiproject is causing a confusing override.

This does need to be fixed, to unblock kim (workaround is admin/admin) and because it's a lousy bug to let through.

**Comment by grechaw** [comment]
I think I've found a small bug in ml-app-deployer.  It will take me a bit to report properly.

The workaround that creates successful deployment is to add to your gradle.properties file

mlManagePassword=YOURADMINPASSWORD

Without this line, the manage client uses 'admin' as a default password for mlDeploy tasks.

I can add this line to the gradle.properties as a workaround as well, but should probably be fixed in ml-app-deployer and then we can upgrade.  I'll check whether we're using the most current ml-gradle anyhow, too.

**Comment by grechaw** [comment]
I spoke too soon.  This change got initial deployment completed, but I just got another 401 further along.

**Comment by grechaw** [comment]
In this case the 401 is in the loadSchemas command.

**Comment by kcoleman-marklogic** [comment]
If you think it would help for me to change my admin password in the parent properties file, I'm perfectly happy to do so as a workaround.

**Comment by kcoleman-marklogic** [comment]
(And I'm not blocked on anything except poking and prodding at your examples. This isn't preventing me from working on the documentation.)

**Comment by grechaw** [comment]
OK I've found where I left off here.  The first workaround appears to be to add to your properties file a new property

mlManagePassword=xxxx

with your admin password in it.   But that's not the entire story, it just gets me most of the way.
It's worth changing the admin username too -- this seems to be an altogether too pervasive kind of bug to be hitting late in the game.

**Comment by grechaw** [comment]
I think the second bug is in the code I wrote into ml-app-deployer for loading the schemas database, a similar but separate defaulting error.

**Comment by grechaw** [comment]
Here is the issue:

```
    /**
     * Like newDatabaseClient, but connects to schemas database.
     *
     * @return
     */
    public DatabaseClient newSchemasDatabaseClient() {
		return DatabaseClientFactory.newClient(getHost(), getRestPort(), getSchemasDatabaseName(),
			getRestAdminUsername(), getRestAdminPassword(), getRestAuthentication(), getRestSslContext(),
			getRestSslHostnameVerifier());
	}
```

The RestAdminPassword, like ManagePassword, is either explicit or defaulted to 'admin'.  It doesn't follow a properties chain.  I can see @rjrudin defending that position.  

What we need to make explicit perhaps is that you need properties for all of the roles, depending on what you expect mldeploy to try to do.

Another reason to merge manage admin and rest admin maybe.

**Comment by grechaw** [comment]
adding mlRestAdminPassword helped here.

So I'm checking the non-admin username case, and will put the full inventory of properties into the examples file.  No reason to confuse or hamper with lack of information in this case!

**Comment by grechaw** [comment]
OK I've added a commit to an existing PR that solves this issue.  We have to decide to triage, but this is a safe and good addition.
Note @jmakeig that I've taken the low road and provided the ubiquitous 'admin/admin' in the examples for every password and username, to underscore just how examply this example is.
Because anyone who has changed the values will know what to do here, I think.

**Comment by jmakeig** [comment]
> I've taken the low road and provided the ubiquitous 'admin/admin' in the examples for every password and username, to underscore just how examply this example is.

Sigh.

**Comment by kcoleman-marklogic** [comment]
Didn't the examples `gradle.properties` file have RestAdminPassword and ManageAdminPassword in it at the beginning of this exercise? Changing the password in those vars didn't help back when I initially reported this.

**Comment by grechaw** [comment]
I think this is ready for you now @kcoleman-marklogic .  You'll see where you have to update username and password pretty explicitly in gradle.properties now.

**Comment by kcoleman-marklogic** [comment]
I am able to successfully deploy the example now. Thank you.
